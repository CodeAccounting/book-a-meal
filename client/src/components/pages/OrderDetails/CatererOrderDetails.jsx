import React, { Component, Fragment } from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import PropTypes from 'prop-types';
import LinkBtn from '../../shared/Link';
import MiniPreloader from '../../shared/Preloader/MiniPreloader';
import View from '../../shared/View';
import { userPropTypes, catererOrderObjPropTypes } from '../../../helpers/proptypes';
import OrderSummary from '../../shared/OrderSummary';
import OrderAmount from '../../shared/OrderAmount';

/**
 * @exports
 * @class CatererOrderDetails
 * @extends Component
 * @classdesc Creates CatererOrderDetails Component
 * @returns {JSX} CatererOrderDetails Component
 */
class CatererOrderDetails extends Component {
  static propTypes = {
    ...userPropTypes,
    order: catererOrderObjPropTypes,
    isFetching: PropTypes.bool.isRequired,
    delivering: PropTypes.bool.isRequired,
    logout: PropTypes.func.isRequired,
    fetchOrder: PropTypes.func.isRequired,
    deliverOrder: PropTypes.func.isRequired,
  }

  static defaultProps = {
    order: null
  }

  /**
   * @memberof CatererOrderDetails
   * @returns {JSX} CatererOrderDetails Component
  */
  componentDidMount() {
    this.props.fetchOrder(this.props.match.params.id);
  }

  /**
   * @memberof CatererOrderDetails
   * @returns {JSX} CatererOrderDetails Component
  */
  deliverOrder = () => {
    this.props.deliverOrder(this.props.order.id);
  }

  /**
   * @memberof CatererOrderDetails
   * @param {object} order
   * @param {func} deliverOrder
   * @returns {JSX} CatererOrderDetails Component
   */
  renderOrderMisc = order => (
    <div className="order-misc">
      <p className="order-date">Order Date: {moment(order.createdAt).format('YY/MM/DD hh:mm')}</p>
      <p>
        Status:&nbsp;
        {order.status === 'canceled' && <span className="danger">Canceled</span>}
        {order.meals[0].delivered && <span className="success">Delivered</span>}
        {!order.meals[0].delivered && order.status !== 'canceled' &&
          <Fragment>
            <LinkBtn className="warning" clickHandler={this.deliverOrder}>Deliver</LinkBtn>
          </Fragment>}
      </p>
      <p>Customer: {`${order.customer.firstname} ${order.customer.lastname}`}</p>
      <p>Address Provided: {order.deliveryAddress}</p>
      <p>Number Provided: {order.deliveryPhoneNo}</p>
    </div>
  )

  /**
   * @memberof CatererOrderDetails
   * @returns {JSX} CatererOrderDetails Component
   */
  renderDetails = () => {
    const { order } = this.props;

    return (
      <Fragment>
        {this.renderOrderMisc(order)}
        <OrderSummary meals={order.meals} status={order.status} />
        <OrderAmount meals={order.meals} type="admin" />
      </Fragment>
    );
  }

  /**
   * @memberof CatererOrderDetails
   * @returns {JSX} CatererOrderDetails Component
   */
  renderCatererOrderDetails = () => {
    const { order } = this.props;

    return (
      <div className="main-wrapper">
        <Link href="/orders" to="/orders" className="orders-back-link">&#8592; Back To Orders</Link>
        <div className="order-confirmation order-details">
          <h3 className="text-center" style={{ paddingTop: '1em' }}>Order #{order.id}</h3>
          {!this.props.delivering && this.renderDetails()}
          {this.props.delivering && <div className="text-center"><MiniPreloader /></div>}
        </div>
      </div>
    );
  }

  /**
   * @memberof CatererOrderDetails
   * @returns {JSX} CatererOrderDetails Component
  */
  render() {
    const { user, logout, isFetching } = this.props;

    return (
      <View user={user} logout={logout} type="orders" showTime isFetching={isFetching}>
        <Fragment>
          {!this.props.order && <p className="text-center info">This Order Does Not Exist</p>}
          {this.props.order && this.renderCatererOrderDetails()}
        </Fragment>
      </View>
    );
  }
}

export default CatererOrderDetails;
