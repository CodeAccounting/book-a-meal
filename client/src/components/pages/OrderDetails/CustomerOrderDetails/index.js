import { connect } from 'react-redux';
import { push } from 'react-router-redux';
import { bindActionCreators } from 'redux';
import CustomerOrderDetails from './CustomerOrderDetails';
import { editOrder, cancelOrder } from '../../../../store/operations/orders';
import { fetchOrder } from '../../../../store/operations/singleOrder';
import { logout } from '../../../../store/operations/auth';

const mapStateToProps = state => ({
  isFetching: state.isFetching,
  order: state.singleOrder.item,
});

const mapDispatchToProps = dispatch => bindActionCreators({
  fetchOrder, logout, push, editOrder, cancelOrder
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(CustomerOrderDetails);
