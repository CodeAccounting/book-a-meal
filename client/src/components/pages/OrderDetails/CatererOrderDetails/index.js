import { connect, } from 'react-redux';
import { bindActionCreators } from 'redux';
import CatererOrderDetails from './CatererOrderDetails';
import { fetchOrders, deliverOrder } from '../../../../store/operations/orders';
import { logout } from '../../../../store/operations/auth';
import { getOrderItem } from '../../../../store/selectors/orders';

const mapStateToProps = (state, props) => ({
  isFetching: state.isFetching,
  delivering: state.orders.delivering,
  order: getOrderItem(props.match.params.id, state.orders.items),
});

const mapDispatchToProps = dispatch => bindActionCreators({
  fetchOrders, deliverOrder, logout
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(CatererOrderDetails);
