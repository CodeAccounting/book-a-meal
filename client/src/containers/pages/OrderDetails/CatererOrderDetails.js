import { connect, } from 'react-redux';
import { bindActionCreators } from 'redux';
import CatererOrderDetails from '../../../components/pages/OrderDetails/CatererOrderDetails';
import { deliverOrder } from '../../../actions/orders';
import { fetchOrder } from '../../../actions/singleOrder';
import { logout } from '../../../actions/auth';

const mapStateToProps = state => ({
  isFetching: state.isFetching,
  delivering: state.orders.delivering,
  order: state.singleOrder.item,
});

const mapDispatchToProps = dispatch => bindActionCreators({
  fetchOrder, deliverOrder, logout
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(CatererOrderDetails);
