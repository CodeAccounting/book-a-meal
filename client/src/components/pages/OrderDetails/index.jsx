import React, { Fragment } from 'react';
import CatererOrderDetails from './CatererOrderDetails';
import CustomerOrderDetails from './CustomerOrderDetails';
import { userPropTypes } from '../../../helpers/proptypes';

/**
 * @exports
 * @function OrderDetails
 * @param {object} props
 * @desc Creates OrderDetails Component
 * @returns {JSX} OrderDetails Component
 */
const OrderDetails = (props) => {
  const { user } = props;
  return (
    <Fragment>
      {user.role === 'caterer' &&
        <CatererOrderDetails {...props} />}
      {user.role === 'customer' &&
        <CustomerOrderDetails {...props} />}
    </Fragment>
  );
};


OrderDetails.propTypes = {
  ...userPropTypes
};

export default OrderDetails;
