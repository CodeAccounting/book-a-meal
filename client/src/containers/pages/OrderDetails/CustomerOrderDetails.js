import { connect } from 'react-redux';
import { push } from 'react-router-redux';
import CustomerOrderDetails from '../../../components/pages/OrderDetails/CustomerOrderDetails';
import { editOrder, cancelOrder } from '../../../actions/orders';
import { fetchOrder } from '../../../actions/singleOrder';
import { logout } from '../../../actions/auth';

const mapStateToProps = state => ({
  isFetching: state.isFetching,
  order: state.singleOrder.item,
});

export default connect(mapStateToProps, {
  fetchOrder, logout, push, editOrder, cancelOrder
})(CustomerOrderDetails);
