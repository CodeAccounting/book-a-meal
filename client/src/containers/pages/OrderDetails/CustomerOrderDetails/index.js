import { connect } from 'react-redux';
import { push } from 'react-router-redux';
import { bindActionCreators } from 'redux';
import CustomerOrderDetails from '../../../../components/pages/OrderDetails/CustomerOrderDetails';
import { editOrder, cancelOrder } from '../../../../actions/orders';
import { fetchOrder } from '../../../../actions/singleOrder';
import { logout } from '../../../../actions/auth';

const mapStateToProps = state => ({
  isFetching: state.isFetching,
  order: state.singleOrder.item,
});

const mapDispatchToProps = dispatch => bindActionCreators({
  fetchOrder, logout, push, editOrder, cancelOrder
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(CustomerOrderDetails);
