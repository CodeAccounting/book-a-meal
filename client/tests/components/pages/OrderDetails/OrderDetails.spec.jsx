import React from 'react';
import OrderDetails from '../../../../src/components/pages/OrderDetails';
import { caterer, customer } from '../../../setup/mockData';

const { now } = Date;

describe('OrderDetails', () => {
  beforeAll(() => {
    Date.now = jest.fn(() => new Date().setMilliseconds(0));
  });

  afterAll(() => {
    Date.now = now;
  });

  it('renders correctly: Caterer OrderDetails', () => {
    const shallowWrapper = shallow(<OrderDetails type="caterer" user={caterer} />);

    expect(toJson(shallowWrapper)).toMatchSnapshot();
    expect(shallowWrapper.find('CatererOrderDetails')).toBeTruthy();
  });

  it('renders correctly: Customer OrderDetails', () => {
    const shallowWrapper = shallow(<OrderDetails type="customer" user={customer} />);

    expect(toJson(shallowWrapper)).toMatchSnapshot();
    expect(shallowWrapper.find('CustomerOrderDetails')).toBeTruthy();
  });
});
