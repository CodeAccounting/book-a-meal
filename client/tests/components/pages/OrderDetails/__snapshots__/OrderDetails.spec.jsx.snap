// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderDetails renders correctly: Caterer OrderDetails 1`] = `
<React.Fragment>
  <Connect(CatererOrderDetails)
    type="caterer"
    user={
      Object {
        "businessAddress": "$, Herbert Macaulay Way, Yaba, Lagos",
        "businessName": "Food Circle",
        "businessPhoneNo": "08166557788",
        "email": "food@circle.com",
        "id": "a09a5570-a3b2-4e21-94c3-5cf483dbd1ac",
        "role": "caterer",
      }
    }
  />
</React.Fragment>
`;

exports[`OrderDetails renders correctly: Customer OrderDetails 1`] = `
<React.Fragment>
  <Connect(CustomerOrderDetails)
    type="customer"
    user={
      Object {
        "email": "emiola@olasanmi.com",
        "firstname": "Emiola",
        "id": "a09a5570-a3b2-4e21-94c3-5cf483dbd1ac",
        "lastname": "Olasanmi",
        "role": "customer",
      }
    }
  />
</React.Fragment>
`;
